File: Lecture27/Shopping.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<cstring>
using namespace std;
int col, row;
int dist[26][26];
char a[25][25];
int scr_x, scr_y;
int des_x, des_y;
int dx[] = {0, 0, 1, -1};
int dy[] = {1, -1, 0, 0};
bool IsPossible(int x, int y) {
	return (x >= 0 && x < row && y >= 0 && y < col && a[x][y] != 'X');
}


void Dijkstra() {
	dist[scr_x][scr_y] = 0;
	// Min-heap: {distance, {x, y}}
	priority_queue<pair<int, pair<int, int>>,
	               vector<pair<int, pair<int, int>>>,
	               greater<pair<int, pair<int, int>>>> pq;
	pq.push({0, {scr_x, scr_y}});

	while (!pq.empty()) {
		auto d = pq.top().first;
		auto pos = pq.top().second;
		pq.pop();
		auto cx = pos.first;
		auto cy = pos.second;

		// Skip outdated entries
		if (d > dist[cx][cy]) continue;
		for (int i = 0; i < 4; i++) {
			int nx = cx + dx[i];
			int ny = cy + dy[i];
			if (IsPossible(nx, ny)) {
				int newDist = dist[cx][cy] + (a[cx][cy] - '0');
				if (newDist < dist[nx][ny]) {
					dist[nx][ny] = newDist;
					pq.push({newDist, {nx, ny}});
				}
			}
		}
	}
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cin >> col >> row;
	while (col && row) {
		for (int i = 0; i < row; i++) {
			for (int j = 0; j < col; j++) {
				cin >> a[i][j];
				dist[i][j] = 1e9;
				if (a[i][j] == 'S') {
					scr_x = i; scr_y = j;
					a[i][j] = '0';
				} else if (a[i][j] == 'D') {
					des_x = i; des_y = j;
					a[i][j] = '0';
				}
			}
		}
		Dijkstra();
		cout << dist[des_x][des_y] << '\n';
		cin >> col >> row;
	}
}

File: Lecture27/Shopping.java
import java.io.*;
import java.util.*;
public class Main {
    static final int[] dx = {0, 0, 1, -1};
    static final int[] dy = {1, -1, 0, 0};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = br.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) continue;
            StringTokenizer st = new StringTokenizer(line);
            int col = Integer.parseInt(st.nextToken());
            int row = Integer.parseInt(st.nextToken());
            if (col == 0 && row == 0) break;
            char[][] a = new char[row][col];
            int[][] dist = new int[row][col];
            int srcX = 0, srcY = 0, desX = 0, desY = 0;
            for (int i = 0; i < row; i++) {
                String s = br.readLine().trim();
                for (int j = 0; j < col; j++) {
                    a[i][j] = s.charAt(j);
                    dist[i][j] = Integer.MAX_VALUE;
                    if (a[i][j] == 'S') {
                        srcX = i; srcY = j;
                        a[i][j] = '0'; // S has 0 cost
                    } else if (a[i][j] == 'D') {
                        desX = i; desY = j;
                        a[i][j] = '0'; // D has 0 cost
                    }
                }
            }
            // Dijkstra with min-heap: {distance, x, y}
            dist[srcX][srcY] = 0;
            PriorityQueue<int[]> pq = new PriorityQueue<>((p, q) -> p[0] - q[0]);
            pq.offer(new int[] {0, srcX, srcY});
            while (!pq.isEmpty()) {
                int[] curr = pq.poll();
                int d = curr[0], cx = curr[1], cy = curr[2];
                // Skip outdated entries
                if (d > dist[cx][cy]) continue;
                for (int i = 0; i < 4; i++) {
                    int nx = cx + dx[i];
                    int ny = cy + dy[i];
                    if (nx >= 0 && nx < row && ny >= 0 && ny < col && a[nx][ny] != 'X') {
                        int newDist = dist[cx][cy] + (a[cx][cy] - '0');
                        if (newDist < dist[nx][ny]) {
                            dist[nx][ny] = newDist;
                            pq.offer(new int[] {newDist, nx, ny});
                        }
                    }
                }
            }
            sb.append(dist[desX][desY]).append('\n');
        }
        System.out.print(sb);
    }
}
